atlasdb-commons/src/test/java/com/palantir/util/SafeShutdownRunnerTest.java

```diff
@@ -0,0 +1,128 @@
+
+    private Runnable mockRunnable = mock(Runnable.class);
+    private Runnable throwingRunnable = mock(Runnable.class);
+    private Runnable verySlowUninterruptibleRunnable = mock(Runnable.class);
+
+    @Before
+    public void setupMocks() {
+        doAnswer(invocation -> {
+            Uninterruptibles.sleepUninterruptibly(10_000, TimeUnit.MILLISECONDS);
+            return null;
+        }).when(verySlowUninterruptibleRunnable).run();
+        doThrow(EXCEPTION).when(throwingRunnable).run();
+    }
+
+    @Test
+    public void runnerRunsOneTask() {
+        SafeShutdownRunner runner = new SafeShutdownRunner(Duration.ofSeconds(1));
+
+        runner.shutdownSafely(mockRunnable);
+
+        verify(mockRunnable, times(1)).run();
+        assertThatCode(runner::close).doesNotThrowAnyException();
+    }
+
+    @Test
+    public void exceptionsAreSuppressedAndReportedWhenClosing() {
+        SafeShutdownRunner runner = new SafeShutdownRunner(Duration.ofSeconds(1));
+
+        assertThatCode(() -> runner.shutdownSafely(throwingRunnable)).doesNotThrowAnyException();
+        assertThatThrownBy(runner::close)
+                .isInstanceOf(RuntimeException.class)
+                .hasSuppressedException(EXCEPTION);
+    }
+
+    @Test
+    public void slowTasksTimeOutWithoutThrowing() {
+        SafeShutdownRunner runner = new SafeShutdownRunner(Duration.ofMillis(50));
+
+        runner.shutdownSafely(verySlowUninterruptibleRunnable);
+        runner.shutdownSafely(verySlowUninterruptibleRunnable);
+
+        verify(verySlowUninterruptibleRunnable, times(2)).run();
+
+        closeAndAssertNumberOfTimeouts(runner, 2);
+    }
+
+    @Test
+    public void otherTasksStillRunAfterSlowTasksThatTimeOut() {
+        SafeShutdownRunner runner = new SafeShutdownRunner(Duration.ofMillis(50));
+
+        runner.shutdownSafely(verySlowUninterruptibleRunnable);
+        runner.shutdownSafely(verySlowUninterruptibleRunnable);
+        runner.shutdownSafely(mockRunnable);
+
+        verify(verySlowUninterruptibleRunnable, times(2)).run();
+        verify(mockRunnable, times(1)).run();
+
+        closeAndAssertNumberOfTimeouts(runner, 2);
+    }
+
+    @Test
+    public void neverRunsRemainingTasksIfExecutorFillsUpWithUninteruptibleTasks() {
+        SafeShutdownRunner runner = new SafeShutdownRunner(Duration.ofMillis(50));
+
+        runner.shutdownSafely(verySlowUninterruptibleRunnable);
+        runner.shutdownSafely(verySlowUninterruptibleRunnable);
+        runner.shutdownSafely(verySlowUninterruptibleRunnable);
+        runner.shutdownSafely(verySlowUninterruptibleRunnable);
+        runner.shutdownSafely(verySlowUninterruptibleRunnable);
+        runner.shutdownSafely(verySlowUninterruptibleRunnable);
+        runner.shutdownSafely(mockRunnable);
+
+        verify(verySlowUninterruptibleRunnable, times(5)).run();
+        verify(mockRunnable, never()).run();
+
+        closeAndAssertNumberOfTimeouts(runner, 7);
```

Comment:
Interesting. I guess it never gets executed, and we drop out after the time limit.

Reply:
Once all the threads are stuck in uninterruptible tasks, everything gets put on the queue and never gets to run as we time out waiting to pick it up from the queue

Reply:
This is no longer applicable with the cached thread pool